internal/providers/pke/pkeworkflow/update_cluster.go

```diff
@@ -83,92 +83,128 @@ func UpdateClusterWorkflow(ctx workflow.Context, input UpdateClusterWorkflowInpu
+		futures := make([]workflow.Future, len(input.NodePoolsToUpdate))
+
+		for i, np := range input.NodePoolsToUpdate {
+			activityInput := UpdatePoolActivityInput{
 				AWSActivityInput: awsActivityInput,
 				Pool:             np,
-				AutoScalingGroup: asgIDs[np.Name],
-			})
-	}
-	for name, future := range futures {
-		err = errors.Append(err, errors.Wrapf(future.Get(ctx, nil), "couldn't update nodepool %q", name))
-	}
-	if err != nil {
-		return err
-	}
+				AutoScalingGroup: asgIDs[i],
+			}
 
-	futures = make(map[string]workflow.Future, len(input.NodePoolsToAdd))
-	for _, np := range input.NodePoolsToAdd {
-		createWorkerPoolActivityInput := CreateWorkerPoolActivityInput{
-			// AWSActivityInput:      awsActivityInput,
-			ClusterID:                 input.ClusterID,
-			Pool:                      np,
-			WorkerInstanceProfile:     PkeGlobalStackName + "-worker-profile",
-			VPCID:                     input.VPCID,
-			VPCDefaultSecurityGroupID: vpcDefaultSecurityGroupID,
-			SubnetID:                  input.SubnetIDs[0],
-			ClusterSecurityGroup:      clusterSecurityGroup,
-			ExternalBaseUrl:           input.PipelineExternalURL,
-			ExternalBaseUrlInsecure:   input.PipelineExternalURLInsecure,
-			SSHKeyName:                "pke-ssh-" + input.ClusterName,
+			futures[i] = workflow.ExecuteActivity(ctx, UpdatePoolActivityName, activityInput)
+		}
+
+		errs := make([]error, len(futures))
+		for i, future := range futures {
+			errs[i] = errors.Wrapf(future.Get(ctx, nil), "couldn't update node pool %q", input.NodePoolsToUpdate[i].Name)
+		}
+
+		if err := errors.Combine(errs...); err != nil {
+			return err
 		}
-		futures[np.Name] = workflow.ExecuteActivity(ctx, CreateWorkerPoolActivityName, createWorkerPoolActivityInput)
 	}
 
-	for name, future := range futures {
-		err = errors.Append(err, errors.Wrapf(future.Get(ctx, nil), "couldn't create nodepool %q", name))
+	{
+		futures := make([]workflow.Future, len(input.NodePoolsToAdd))
+
+		for i, np := range input.NodePoolsToAdd {
+			createWorkerPoolActivityInput := CreateWorkerPoolActivityInput{
+				// AWSActivityInput:      awsActivityInput,
+				ClusterID:                 input.ClusterID,
+				Pool:                      np,
+				WorkerInstanceProfile:     PkeGlobalStackName + "-worker-profile",
+				VPCID:                     input.VPCID,
+				VPCDefaultSecurityGroupID: vpcDefaultSecurityGroupID,
+				SubnetID:                  input.SubnetIDs[0],
+				ClusterSecurityGroup:      clusterSecurityGroup,
+				ExternalBaseUrl:           input.PipelineExternalURL,
+				ExternalBaseUrlInsecure:   input.PipelineExternalURLInsecure,
+				SSHKeyName:                "pke-ssh-" + input.ClusterName,
+			}
+
+			futures[i] = workflow.ExecuteActivity(ctx, CreateWorkerPoolActivityName, createWorkerPoolActivityInput)
+		}
+
+		errs := make([]error, len(futures))
+		for i, future := range futures {
+			errs[i] = errors.Wrapf(future.Get(ctx, nil), "couldn't create node pool %q", input.NodePoolsToAdd[i].Name)
```

Comment:
to be honest, I don't feel this approach better than the append

Reply:
Append dynamically resizes the underlying array, while this approach does not require that.

Reply:
Not necessarily. `append` only resizes the slice when its capacity would be exceeded. If the initial capacity is chosen well, no resizing will happen. Also, append will fill the slice without leaving "gaps", while explicit indexed assignment might leave gaps when non-assigning branches are taken (see the nil ptr problem above).

Reply:
In this case not leaving gaps is not an option, because the index is used to access elements in the original slice.